### PR TITLE
feat: add OptimismNode type

### PR DESCRIPTION
**Pull request by mattsse**
more builder prep

### CODE REVIEW

File: crates/node-optimism/src/node.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+                |shutdown| {
+                    reth_transaction_pool::maintain::backup_local_transactions_task(
+                        shutdown,
+                        pool.clone(),
+                        transactions_backup_config,
+                    )
+                },
+            );
+
+            // spawn the maintenance task
```
Comment (emhane) [inline crates/node-optimism/src/node.rs:104]: maintenance gives the notion it's a background task that is periodically doing some optimisation task. for for example reordering elements in some store or moving elements between two different stores.

File: crates/node-optimism/src/node.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        }
+
+        Ok(transaction_pool)
+    }
+}
+
+/// A basic ethereum payload service.
+#[derive(Debug, Default, Clone)]
+#[non_exhaustive]
+pub struct EthereumPayloadBuilder;
```
Comment (emhane) [inline crates/node-optimism/src/node.rs:125]: what do you actually mean by payload builder? it builds transactions? and it's spawned in a thread, where does it return the payload once it's built, via a callback to the initiator of the job? or is sending a command to this service the entry point to the p2p? some more docs would be good.

**Review by emhane — CHANGES_REQUESTED** [review]
components are all services, between which components is control flow based on communication (message passing model) or locks (shared memory model)? is control centralised from the type that impl engine trait, or which components can work together without the engine? which component depends on which, can the components in ATs in `NodeTypes` be run on their own or are they all interdependent? do they need to be started in a particular order? some more docs on control flow, coupling (and cohesion) would be nice.

some more docs on how components communicate with each other and with the network for components that communicate with the network, would be very helpful, e.g. is the communication uni-directional or bidirectional, is the flow of messages a DAG or cyclic? what types of messages are being sent/received, on which channels/sockets?

File: crates/node-optimism/src/node.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
```
Comment (Rjected) [inline crates/node-optimism/src/node.rs:125]: why is this called `EthereumPayloadBuilder`, but then it creates the optimism payload builder in `spawn_payload_service`?

**Review by Rjected — COMMENTED** [review]
one question about naming

File: crates/node-optimism/src/node.rs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+where
+    Node: FullNodeTypes<Engine = OptimismEngineTypes>,
+    Pool: TransactionPool + Unpin + 'static,
+{
+    fn spawn_payload_service(
+        self,
+        ctx: &BuilderContext<Node>,
+        pool: Pool,
+    ) -> eyre::Result<PayloadBuilderHandle<Node::Engine>> {
+        let payload_builder = reth_optimism_payload_builder::OptimismPayloadBuilder::default();
```
Comment (Rjected) [inline crates/node-optimism/src/node.rs:136]: do we need to define `compute_pending_block` in the above `OptimismPayloadBuilder`?

**Review by Rjected — APPROVED** [review]
lgtm so far, pending one question about `compute_pending_block`. I also wonder if we can just consolidate the `reth_optimism_payload_builder::OptimismPayloadBuilder` and `reth_node_builder::OptimismPayloadBuilder` types. With these new traits and types we're seeing a lot of overloaded names, it's worth thinking about and improving in followups

File: crates/node-optimism/src/node.rs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+        // spawn txpool maintenance task
+        {
+            let pool = transaction_pool.clone();
+            let chain_events = ctx.provider().canonical_state_stream();
+            let client = ctx.provider().clone();
+            let transactions_backup_config =
+                reth_transaction_pool::maintain::LocalTransactionBackupConfig::with_local_txs_backup(transactions_path);
+
+            ctx.task_executor().spawn_critical_with_graceful_shutdown_signal(
+                "local transactions backup task",
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/node-optimism/src/node.rs:93]: @emhane this backsup local transactions on disk so we can recover them when we restart the node

File: crates/node-optimism/src/node.rs (hunk 2 above)
Comment (mattsse) [reply crates/node-optimism/src/node.rs]: this sits behind the engine API, if the CL tells us to build a payload (block) this type is responsible for building one and serves it to the CL if it requests the payload

the default impl will use the mempool to build a new payload

File: crates/node-optimism/src/node.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (emhane) [reply crates/node-optimism/src/node.rs:93]: ok cool, please add a doc that I has to do with backing up txns to disk

File: crates/node-optimism/src/node.rs (hunk 2 above)
Comment (emhane) [reply crates/node-optimism/src/node.rs]: please add to docs

**Pull request merged**